src/Microsoft.AspNetCore.Hosting/WebHostBuilder.cs

```diff
@@ -225,8 +180,50 @@ private IServiceCollection BuildHostingServices()
             if (!string.IsNullOrEmpty(_options.ServerAssembly))
             {
                 // Add the server
-                var serverType = ServerLoader.ResolveServerType(_options.ServerAssembly);
-                services.AddSingleton(typeof(IServer), serverType);
+                try
+                {
+                    var serverType = ServerLoader.ResolveServerType(_options.ServerAssembly);
+                    services.AddSingleton(typeof(IServer), serverType);
+                }
+                catch (Exception ex)
+                {
+                    var capture = ExceptionDispatchInfo.Capture(ex);
+                    services.AddSingleton<IServer>(_ =>
+                    {
+                        capture.Throw();
+                        return null;
+                    });
+                }
+            }
+
+            if (!string.IsNullOrEmpty(_options.StartupAssembly))
+            {
+                try
+                {
+                    var startupType = StartupLoader.FindStartupType(_options.StartupAssembly, _hostingEnvironment.EnvironmentName);
+                    
+                    if (typeof(IStartup).GetTypeInfo().IsAssignableFrom(startupType.GetTypeInfo()))
+                    {
+                        services.AddSingleton(typeof(IStartup), startupType);
```

Comment:
Hmm, this keeps the startup instance alive for the duration of the app. It also would call Dispose at the end if implemented.